### PR TITLE
LPD-31927 Support for SKU's ERC when creating a Product Option Value via Headless APIs

### PR DESCRIPTION
**Pull request by monaco-fabio**
https://liferay.atlassian.net/browse/LPD-31927

### CODE REVIEW

**Comment by monaco-fabio** [comment]
ci:test:sf

**Comment by monaco-fabio** [comment]
ci:test:relevant

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java [hunk 1]
```diff
@@ -104,9 +104,15 @@ public ProductOptionValue patchProductOptionValue(
 
 		long cpInstanceId = 0;
 
-		CPInstance cpInstance = _cpInstanceService.fetchCProductInstance(
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java:107]: I think this should not be removed, what id the user don't pass both ERC and SKUId? PATCH should not erase the previous SKUID

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java [hunk 2]
```diff
@@ -104,9 +104,15 @@ public ProductOptionValue patchProductOptionValue(
...
-			cpDefinitionOptionValueRel.getCPInstanceUuid());
+		CPInstance cpInstance =
+			_cpInstanceLocalService.fetchByExternalReferenceCode(
+				productOptionValue.getSkuExternalReferenceCode(),
+				contextCompany.getCompanyId());
+
+		if (cpInstance == null) {
+			cpInstance = _cpInstanceLocalService.fetchCPInstance(
+				GetterUtil.getLong(productOptionValue.getSkuId()));
+		}
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java:115]: just add a null check here

Comment (andrea-ale-sbarra) [reply modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java:115]:
```suggestion
		}
		if (cpInstance==null) {
		CPInstance cpInstance = _cpInstanceService.fetchCProductInstance(
		cpDefinitionOptionValueRel.getCProductId(),
		cpDefinitionOptionValueRel.getCPInstanceUuid());
		}
```

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/util/v1_0/ProductOptionValueUtil.java [hunk 3]
```diff
@@ -37,13 +38,17 @@ public class ProductOptionValueUtil {
 			cpDefinitionOptionValueRelService.fetchCPDefinitionOptionValueRel(
 				cpDefinitionOptionRelId, productOptionValue.getKey());
 
-		CPInstance cpInstance = null;
+		long cpInstanceId = 0;
 
-		if (cpDefinitionOptionValueRel != null) {
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/util/v1_0/ProductOptionValueUtil.java:42]: also here why did you remove this piece of code?

**Review by andrea-ale-sbarra — CHANGES_REQUESTED** [review]
Please have a look at my comments.

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java [hunk 4]
```diff
@@ -188,7 +194,7 @@ private ProductOptionValue _toProductOptionValue(
 		_cpDefinitionOptionValueRelService;
 
 	@Reference
-	private CPInstanceService _cpInstanceService;
+	private CPInstanceLocalService _cpInstanceLocalService;
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-impl/src/main/java/com/liferay/headless/commerce/admin/catalog/internal/resource/v1_0/ProductOptionValueResourceImpl.java:197]: Use the remote service (to check if you have the permission to perform the action)

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 5]
```diff
@@ -6,15 +6,331 @@
...
+		_commerceCatalog = CommerceTestUtil.addCommerceCatalog(
+			testCompany.getCompanyId(), testCompany.getGroupId(),
+			_user.getUserId(), _commerceCurrency.getCode());
+
+		_cpInstance = CPTestUtil.addCPInstanceWithRandomSku(
+			testGroup.getGroupId(), BigDecimal.TEN);
+
+		_cpDefinition = CPTestUtil.addCPDefinitionFromCatalog(
+			_commerceCatalog.getGroupId(), SimpleCPTypeConstants.NAME, false,
+			false);
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:67]: the _cpDefinition should not be the one related to the _cpInstance created?

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 6]
```diff
@@ -6,15 +6,331 @@
...
+
+		ServiceContext serviceContext =
+			ServiceContextTestUtil.getServiceContext(
+				testCompany.getCompanyId(), testGroup.getGroupId(),
+				_user.getUserId());
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:72]: inline this where used (I think only creating the _cpDefinitionOptionRel)

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 7]
```diff
@@ -6,15 +6,331 @@
...
+
+		BeanTestUtil.copyProperties(
+			randomPatchProductOptionValue, expectedPatchProductOptionValue);
+
+		ProductOptionValue getProductOptionValue =
+			productOptionValueResource.getProductOptionValue(
+				patchProductOptionValue.getId());
+
+		assertEquals(expectedPatchProductOptionValue, getProductOptionValue);
+		assertValid(getProductOptionValue);
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:163]: what are the differences compared with the test in the base class?

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 8]
```diff
@@ -6,15 +6,331 @@
...
+
+		_testPatchProductOptionIdProductOptionValueWithSkuExternalReferenceCode();
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:165]: this should be _testPatchProductOptionValueWith**** as inside this test

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 9]
```diff
@@ -6,15 +6,331 @@
...
+	public void testPostProductOptionIdProductOptionValue() throws Exception {
+		ProductOptionValue randomProductOptionValue =
+			randomProductOptionValue();
+
+		ProductOptionValue postProductOptionValue =
+			testPostProductOptionIdProductOptionValue_addProductOptionValue(
+				randomProductOptionValue);
+
+		assertEquals(randomProductOptionValue, postProductOptionValue);
+		assertValid(postProductOptionValue);
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:179]: differences compared with the base one?

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 10]
```diff
@@ -6,15 +6,331 @@
...
+		ProductOptionValue randomPatchProductOptionValue =
+			randomProductOptionValue();
+
+		CPInstance cpInstance = CPTestUtil.addCPInstanceWithRandomSku(
+			testGroup.getGroupId(), BigDecimal.TEN);
+
+		randomPatchProductOptionValue.setSkuId(0L);
+
+		randomPatchProductOptionValue.setSkuExternalReferenceCode(
+			cpInstance.getExternalReferenceCode());
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:260]: swipe

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 11]
```diff
@@ -6,15 +6,331 @@
...
+		BeanTestUtil.copyProperties(
+			randomPatchProductOptionValue, expectedPatchProductOptionValue);
+
+		ProductOptionValue getProductOptionValue =
+			productOptionValueResource.getProductOptionValue(
+				patchProductOptionValue.getId());
+
+		assertEquals(expectedPatchProductOptionValue, getProductOptionValue);
+		assertValid(getProductOptionValue);
+	}
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:278]: please add the check if the two ids are the same and the two ercs are the same

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 12]
```diff
@@ -6,15 +6,331 @@
...
+		CPInstance cpInstance = CPTestUtil.addCPInstanceWithRandomSku(
+			testGroup.getGroupId(), BigDecimal.TEN);
+
+		long cpInstanceId = cpInstance.getCPInstanceId();
+
+		randomProductOptionValue.setSkuId(0L);
+
+		randomProductOptionValue.setSkuExternalReferenceCode(
+			cpInstance.getExternalReferenceCode());
+
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:295]: swipe

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 13]
```diff
@@ -6,15 +6,331 @@
...
+		randomProductOptionValue.setSkuId(cpInstanceId);
+
+		assertEquals(randomProductOptionValue, postProductOptionValue);
+		assertValid(postProductOptionValue);
+		Assert.assertEquals(
+			cpInstanceId,
+			GetterUtil.getLong(postProductOptionValue.getSkuId()));
+		Assert.assertEquals(
+			randomProductOptionValue.getSkuExternalReferenceCode(),
+			postProductOptionValue.getSkuExternalReferenceCode());
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:309]: swipe them

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java (hunk 11 above)
Comment (smottal) [reply modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:278]: like you did in _testPostProductOptionIdProductOptionValueWithSkuExternalReferenceCode

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 14]
```diff
@@ -6,15 +6,331 @@
...
+	}
+
+	@DeleteAfterTestRun
+	private CommerceCatalog _commerceCatalog;
+
+	private CommerceCurrency _commerceCurrency;
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:315]: @DeleteAfterTestRun

**Review by smottal — CHANGES_REQUESTED** [review]

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 15]
```diff
@@ -6,15 +6,311 @@
...
+		_user = UserTestUtil.addUser(testCompany);
+
+		_commerceCatalog = CommerceTestUtil.addCommerceCatalog(
+			testCompany.getCompanyId(), testCompany.getGroupId(),
+			_user.getUserId(), _commerceCurrency.getCode());
+
+		_cpInstance = CPTestUtil.addCPInstanceWithRandomSku(
+			testGroup.getGroupId(), BigDecimal.TEN);
+
+		_cpDefinition = CPTestUtil.addCPDefinitionFromCatalog(
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:64]: move this before the _cpInstance

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java [hunk 16]
```diff
@@ -6,15 +6,311 @@
...
+		BeanTestUtil.copyProperties(
+			randomPatchProductOptionValue, expectedPatchProductOptionValue);
+
+		ProductOptionValue getProductOptionValue =
+			productOptionValueResource.getProductOptionValue(
+				patchProductOptionValue.getId());
+
+		assertEquals(expectedPatchProductOptionValue, getProductOptionValue);
+		assertValid(getProductOptionValue);
+
```
Comment (smottal) [inline modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:250]: remove this empty line

**Review by smottal — CHANGES_REQUESTED** [review]

File: modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java (hunk 16 above)
Comment (monaco-fabio) [reply modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-catalog-test/src/testIntegration/java/com/liferay/headless/commerce/admin/catalog/resource/v1_0/test/ProductOptionValueResourceTest.java:247]: Format source want that empty line, if i remove the line, format fails

**Comment by monaco-fabio** [comment]
ci:test:sf

**Comment by monaco-fabio** [comment]
ci:test:sf

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
ci:test:relevant

**Comment by monaco-fabio** [comment]
ci:test:sf

**Comment by monaco-fabio** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:forward:force
